src/nvim/api/buffer.c

```diff
@@ -725,11 +727,47 @@ void nvim_buf_set_text(uint64_t channel_id, Buffer buffer, Integer start_row, In
 
   FOR_ALL_TAB_WINDOWS(tp, win) {
     if (win->w_buffer == buf) {
-      // adjust cursor like an extmark ( i e it was inside last_part_len)
-      if (win->w_cursor.lnum == end_row && win->w_cursor.col > end_col) {
-        win->w_cursor.col -= col_extent - (colnr_T)last_item.size;
+      if (win->w_cursor.lnum >= start_row && win->w_cursor.lnum <= end_row) {
+        colnr_T mode_col_adj = win == curwin && (State & MODE_INSERT) ? 0 : 1;
+        colnr_T cursor_gravity_col = win->w_cursor.col + mode_col_adj;
+
+        if (win->w_cursor.lnum == end_row && cursor_gravity_col > end_col) {
+          win->w_cursor.lnum += (linenr_T)extra;
+          win->w_cursor.col += (colnr_T)last_item.size - (colnr_T)end_col;
+
+          if (new_len == 1) {
+            win->w_cursor.col += (colnr_T)start_col;
+          }
+        } else {
+          linenr_T repl_end_row = (linenr_T)start_row + (linenr_T)replacement.size - 1;
+          bool cursor_was_after_end_row = win->w_cursor.lnum > repl_end_row;
+
+          if (cursor_was_after_end_row) {
+            win->w_cursor.lnum = repl_end_row;
+          }
+
+          colnr_T repl_start_col = new_len == 1 ? (colnr_T)start_col : 0;
+          int index = win->w_cursor.lnum - (int)start_row;
+          String repl_item = replacement.items[index].data.string;
+          colnr_T repl_end_col = repl_start_col + (colnr_T)repl_item.size;
+
+          if (cursor_was_after_end_row
+              || (win->w_cursor.lnum == repl_end_row
+                  && win->w_cursor.col > repl_end_col)) {
+            win->w_cursor.col = repl_end_col;
+
+            if (win->w_cursor.col - mode_col_adj >= repl_start_col) {
+              win->w_cursor.col -= mode_col_adj;
+            }
+          }
+        }
+
+        check_cursor_col_win(win);
+        changed_cline_bef_curs(win);
+        invalidate_botline(win);
```

Comment:
I'd suggest to break out this into a new `fix_cursor_cols` function, even if it is just called in one place.